plugins/parsers/parquet/parser.go

```diff
@@ -0,0 +1,152 @@
+package parquet
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/apache/arrow/go/v16/parquet/file"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+type Parser struct {
+	MeasurementColumn string   `toml:"measurement_column"`
+	TagColumns        []string `toml:"tag_columns"`
+	TimestampColumn   string   `toml:"timestamp_column"`
+	TimestampFormat   string   `toml:"timestamp_format"`
+	TimestampTimezone string   `toml:"timestamp_timezone"`
+
+	defaultTags map[string]string
+	location    *time.Location
+	metricName  string
+}
+
+func (p *Parser) Init() error {
+	if p.TimestampFormat == "" {
+		p.TimestampFormat = "unix"
+	}
+	if p.TimestampTimezone == "" {
+		p.location = time.UTC
+	} else {
+		loc, err := time.LoadLocation(p.TimestampTimezone)
+		if err != nil {
+			return fmt.Errorf("invalid location %s: %w", p.TimestampTimezone, err)
+		}
+		p.location = loc
+	}
+
+	return nil
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	reader := bytes.NewReader(buf)
+	parquetReader, err := file.NewParquetReader(reader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create parquet reader: %w", err)
+	}
+
+	metadata := parquetReader.MetaData()
+	columns := make([]string, metadata.Schema.NumColumns())
+	data := make([][]any, metadata.Schema.NumColumns())
+	for i := 0; i < parquetReader.NumRowGroups(); i++ {
+		rowGroup := parquetReader.RowGroup(i)
+		scanners := make([]*columnParser, metadata.Schema.NumColumns())
+		for colIndex := range metadata.Schema.NumColumns() {
+			col, err := rowGroup.Column(colIndex)
+			if err != nil {
+				return nil, fmt.Errorf("unable to fetch column %q: %w", colIndex, err)
+			}
+
+			scanners[colIndex] = newColumnParser(col)
+			columns[colIndex] = col.Descriptor().Name()
+		}
+
+		for i, s := range scanners {
+			for s.HasNext() {
+				if val, ok := s.Next(); ok {
+					data[i] = append(data[i], val)
+				}
+			}
+		}
+	}
+
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	metrics := make([]telegraf.Metric, len(data[0]))
```

Comment:
You said columns are not necessarily defined for all rows, so can we be sure that the first column has all rows or do we need to find the "max" length over all rows?

I'm asking because you later check against this index and if the first column has less rows compared to others this might panic...

Reply:
Good catch! Updated.